src/twitch-api.js

```diff
@@ -0,0 +1,171 @@
+   * @type {?ApiClient}
+   */
+  #apiClient;
+  /**
+   * @type {?import('@twurple/api').HelixUser}
+   */
+  #broadcasterUser;
+  /**
+   * @type {import('@twurple/api').HelixChatChatter[]}
+   */
+  #chattersCache = [];
+  /**
+   * @type {Date}
+   */
+  #chattersCacheTime = null;
+
+  /**
+   * Setup authentication.
+   */
+  async setup() {
+    if (
+      settings.clientId == null ||
+      settings.clientId == "" ||
+      settings.clientId == "{YOUR_CLIENT_ID}"
+    ) {
+      throw new Error(`${settings.fileName}: Invalid clientId.`);
+    }
+    if (
+      settings.clientSecret == null ||
+      settings.clientSecret == "" ||
+      settings.clientSecret == "{YOUR_CLIENT_SECRET}"
+    ) {
+      throw new Error(`${settings.fileName}: Invalid clientSecret.`);
+    }
+    // validation is done before setting anything up to help users figure out problems early
+    const tokenData = JSON.parse(fs.readFileSync(tokensFileName, "utf-8"));
+    if (
+      tokenData.accessToken == null ||
+      tokenData.accessToken == "" ||
+      tokenData.accessToken == "{INITIAL_ACCESS_TOKEN}"
+    ) {
+      throw new Error(`Invalid ${tokensFileName} file: accessToken not found.`);
+    }
+    if (
+      tokenData.refreshToken == null ||
+      tokenData.refreshToken == "" ||
+      tokenData.refreshToken == "{INITIAL_REFRESH_TOKEN}"
+    ) {
+      throw new Error(
+        `Invalid ${tokensFileName} file: refreshToken not found.`
+      );
+    }
+    // create the refreshing provider
+    this.#authProvider = new RefreshingAuthProvider({
+      clientId: settings.clientId,
+      clientSecret: settings.clientSecret,
+      onRefresh: async (userId, newTokenData) =>
+        await gracefulFs.writeFile(
+          tokensFileName,
+          JSON.stringify(newTokenData, null, 4),
+          "utf-8"
+        ),
+    });
+    // register refresh and access token of the bot and get the user id of the bot
+    // this token is used for both chat as well as api calls
+    this.#botUserId = await this.#authProvider.addUserForToken(tokenData, [
+      "chat",
+      "user-by-name",
+      "chatters",
+    ]);
+    // create the api client
+    this.#apiClient = new ApiClient({ authProvider: this.#authProvider });
+    // get the user id of the channel/broadcaster
+    this.#broadcasterUser = await this.#apiClient.asIntent(
+      ["user-by-name"],
+      async (ctx) => {
+        return ctx.users.getUserByName(settings.channel);
+      }
+    );
+  }
+
+  /**
+   * This creates a new tmi client where the authentication is setup automatically.
+   *
+   * How the tmi client is authenticating is transparent to the user.
+   *
+   * @param {import('@types/tmi.js').Options} opts The options of the tmi client.
+   * @returns {import('@types/tmi.js').Client} A new tmi client that has authentication build in.
+   */
+  createTmiClient(opts) {
+    return new tmi.client({ ...opts, authProvider: this.#authProvider });
+  }
+
+  async #useCache(forceLoad = false) {
```

Comment:
I implemented this, however I have (4) before (2) and in this case I return the cached value, so i assume the current solution could still be improved by doing step (4) as you suggested instead of how it currently is.

Reply:
Oh, and step (1) and (2) is swapped in my implementation as well.

Reply:
I quite like how you've got it implemented here, actually (particularly the new BroadcastOnce class).

I *did* think there was potentially a bit of a race condition here where a refresh could push us under the rate limit threshold (3), await on getChatters, and then while it's waiting another command comes in and (seeing we're now under the rate limit threshold) resolves all the promises... but that's not actually how the rate limit check works anyway! So it's fine.